archive/tar_stream.ts

```diff
@@ -0,0 +1,339 @@
+}
+
+/**
+ * The interface required to provide a directory.
+ */
+export interface TarStreamDir {
+  pathname: string | [Uint8Array, Uint8Array];
+  options?: Partial<TarStreamOptions>;
+}
+
+/**
+ * A union type merging all the TarStream interfaces that can be piped into the
+ * TarStream class.
+ */
+export type TarStreamInput = TarStreamFile | TarStreamDir;
+
+/**
+ * The options that can go along with a file or directory.
+ * @param mode An octal number in ASCII.
+ * @param uid An octal number in ASCII.
+ * @param gid An octal number in ASCII.
+ * @param mtime A number of seconds since the start of epoch. Avoid negative
+ * values.
+ * @param uname An ASCII string. Should be used in preference of uid.
+ * @param gname An ASCII string. Should be used in preference of gid.
+ * @param devmajor The major number for character device.
+ * @param devminor The minor number for block device entry.
+ */
+export interface TarStreamOptions {
+  mode: string;
+  uid: string;
+  gid: string;
+  mtime: number;
+  uname: string;
+  gname: string;
+  devmajor: string;
+  devminor: string;
+}
+
+const SLASH_CODE_POINT = "/".charCodeAt(0);
+
+/**
+ * ### Overview
+ * A TransformStream to create a tar archive.  Tar archives allow for storing
+ * multiple files in a single file (called an archive, or sometimes a tarball).
+ *   These archives typically have a single '.tar' extension.  This
+ * implementation follows the [FreeBSD 15.0](https://man.freebsd.org/cgi/man.cgi?query=tar&sektion=5&apropos=0&manpath=FreeBSD+15.0-CURRENT) spec.
+ *
+ * ### File Format & Limitations
+ * The ustar file format is used for creating the tar archive.  While this
+ * format is compatible with most tar readers, the format has several
+ * limitations, including:
+ * - Pathnames must be at most 256 characters.
+ * - Files must be at most 8 GiBs in size, or 64 GiBs if `sizeExtension` is set
+ * to true.
+ * - Sparse files are not supported.
+ *
+ * ### Usage
+ * TarStream may throw an error for several reasons. A few of those are:
+ * - The pathname is invalid.
+ * - The size provided does not match that of the iterable's length.
+ *
+ * ### Compression
+ * Tar archives are not compressed by default.  If you'd like to compress the
+ * archive, you may do so by piping it through a compression stream.
+ *
+ * @example
+ * ```ts
+ * import { TarStream } from "@std/archive/tar-stream";
+ *
+ * await ReadableStream.from([
+ *   {
+ *     pathname: 'potato/'
+ *   },
+ *   {
+ *     pathname: 'deno.json',
+ *     size: (await Deno.stat('deno.json')).size,
+ *     iterable: (await Deno.open('deno.json')).readable
+ *   },
+ *   {
+ *     pathname: 'deno.lock',
+ *     size: (await Deno.stat('deno.lock')).size,
+ *     iterable: (await Deno.open('deno.lock')).readable
+ *   }
+ * ])
+ *   .pipeThrough(new TarStream())
+ *   .pipeThrough(new CompressionStream('gzip'))
+ *   .pipeTo((await Deno.create('./out.tar.gz')).writable)
+ * ```
+ */
+export class TarStream {
```

Comment:
While TarStream could technically extend TransformStream, it can easily cause high memory usage and even overflow it causing a crash. The problem with extending TransformStreams here is that for every one TarStreamInput in, the entire files content gets enqueued, and since iterables can be at most 64 GiBs, a slow read out of the TarStream could cause a crash and will cause high memory spikes.

Reply:
nope, it cannot, since a different readable stream is necessary that does some processing. however it should `implement TransformStream`